add custom domain

ann26 opened the pull request:
fix #570 and fix #603 

# [ WIP ]

This PR:
- created custom domain model where subscribed user and user domain is saved
- when the user and domain are approved, it will write the user domain to `server-name.txt` which is included in nginx so the domain will be included in the allowed hosts.

@timlinux and @cchristelis is this the right approach?

in the admin panel:
![screenshot from 2017-11-29 18-03-07](https://user-images.githubusercontent.com/26101337/33372341-405f8136-d530-11e7-8a6b-c219054e7841.png)
![screenshot from 2017-11-29 18-03-20](https://user-images.githubusercontent.com/26101337/33372345-41444f64-d530-11e7-9404-3a256b5571a1.png)

timlinux commented on django_project/base/models/custom_domain.py:
```diff
@@ -0,0 +1,57 @@
+)
+
+
+class CustomDomain(models.Model):
```
I think we should just call it 'Domain' and then create a Domain entry for e.g. http://projecta.kartoza.com which is the default domain that all new projects get added to that do not have an explicitly set domain. Then add a foreign key reference to Domain from both the organisation and project models.

timlinux commented on deployment/sites-enabled/default.conf:
```diff
@@ -20,6 +20,7 @@ server {
     # the domain name it will serve for
     server_name staging.projecta.kartoza.com changelog.linfiniti.com changelog.kartoza.com changelog.qgis.org changelog.inasafe.org staging.changelog.qgis.org;
     charset     utf-8;
+    include /home/web/media/server-name.txt
```
I would imaging that since we will on a dedicated server using * for server name might be a simpler approach and then let projecta handle whether the hostname in the header of the request if valid or not.

timlinux replied:
(that we we do not need to use server-name.txt tricks)

timlinux commented on django_project/base/models/custom_domain.py:
```diff
@@ -0,0 +1,57 @@
+
+    class Meta:
+        ordering = ['user']
+        unique_together = ['user', 'custom_domain']
```
I think the domain would need to be independently unique - otherwise what will happen if two users try to register the same domain?

timlinux commented on django_project/base/models/custom_domain.py:
```diff
@@ -0,0 +1,57 @@
+    )
+
+    class Meta:
+        ordering = ['user']
```
lets order by the domain name

timlinux commented:
@ann26 Referring back to my original ticket at #635 I think it makes more sense to make an atomic PR that adds an organisation model and associates each project with each organisation. By default we can associate projects with the 'global' organisation until the are assigned to a specific organisation.

Then each organisation and project will get associated with a domain.

timlinux commented:
@ann26 can you also take a look at this PR from @meomancer and review whether the code will fit within the framework we are setting up here:

https://github.com/kartoza/projecta/pull/361

ann26 commented:
ok @timlinux will do

ann26 commented:
Hi @timlinux , looking at @meomancer 's PR (#361) I think the code can fit our framework with a bit of modification. 

I tried it in my local and using it we can assign additional administrators that can have permissions the same as the owner in the project, as shown below:
![screenshot from 2017-12-21 20-35-16](https://user-images.githubusercontent.com/26101337/34258092-06da83dc-e68f-11e7-82df-35ffbef6503f.png)
So in above example case, I am (not a staff) the owner of Inasafe project and an administrator of XProject project. I have the same permissions as the owner of XProject eventhough I am not the owner of it. 

The app also includes custom template tag that we can implement in template, e.g. `the_project|is_project_administrator:user` 

Maybe the Project Collaborators can be the Managers of each app (e.g. changelog, sponsorship, certification) ?

In conclusion, I think using permission app, we can manage permissions separately from the project model and its custom template tag can simplify the template code.

ann26 commented:
Hi @timlinux , here is update:
- [x] Added organisation model
- [x] Added organisation create and list view
- [x] The existing projects are assigned to default organisation (Kartoza)
- [x] Added domain model
- [x] Added domain create and list view
- [x] Domain list view only available for staff
- [x] Added domain check middleware
- [x] Changed nginx conf server name to `*` (wildcard)
- [x] When a domain is registered with role as project, the site will point to project page
- [x] When a domain is registered with role as organisation, the site will point to home page (project list page) with only showing the project in that organisation.
- [x] Added domain check middleware test
- [x] Update existing tests

Gifs:
- Organisation create view
![peek 2018-01-03 17-07](https://user-images.githubusercontent.com/26101337/34551015-4b05a9e2-f148-11e7-9305-1f8b5c012d91.gif)

- Organisation list and pending view
![peek 2018-01-03 17-08](https://user-images.githubusercontent.com/26101337/34551025-5d29c860-f148-11e7-9f8f-f3955e3941c2.gif)

- Domain create view
User may click on `register a domain` from navigation.
![peek 2018-01-03 17-32](https://user-images.githubusercontent.com/26101337/34551108-17f01852-f149-11e7-967b-d32c003234e7.gif)

Or may click on `set up custom domain` button on project page
![peek 2018-01-03 17-42](https://user-images.githubusercontent.com/26101337/34551157-95fbe1a4-f149-11e7-8613-39b9ce375c3d.gif)

- When domain is not registered or not approved yet
![peek 2018-01-03 17-35](https://user-images.githubusercontent.com/26101337/34551102-09246422-f149-11e7-87f3-aec87f28bf8a.gif)

- Domain list view (only available to staff users)
![peek 2018-01-03 17-36](https://user-images.githubusercontent.com/26101337/34551124-3a3efd38-f149-11e7-9571-4e2d643ae2a4.gif)

- Domain as organisation role
![peek 2018-01-03 17-38](https://user-images.githubusercontent.com/26101337/34551133-502516aa-f149-11e7-9789-a9886ee0cef0.gif)

- Domain as project role
![peek 2018-01-03 17-40](https://user-images.githubusercontent.com/26101337/34551140-60ffd3b6-f149-11e7-9947-36dd44c25e19.gif)


Travis is still failing, I am still investigating it.

timlinux commented:
Wow @ann26 beautiful work! Please merge when you have travis passing and then deploy it to staging so I can play with it a bit.
